### ISSUE TITLE
Allow users to trigger node drift

### ISSUE THREAD

**Issue by sidewinder12s** [aws/karpenter#4231]
### Description

**What problem are you trying to solve?**

In Karpenter pre-v0.28, I had started using the karpenter drift annotation `karpenter.sh/voluntary-disruption=drifted` as a way to force nodes to get replaced in an orderly fashion when I changed configuration that was not supported by Karpenters Drift detection. 

In v0.28 this was removed and now the annotation is simply removed by Karpenter. 

I found the ability to trigger drift useful in testing and in filling in the gaps in Drift support. I'd also assume long term, there may be corner cases users would want to trigger replacement on that drift cannot detect or detect easily. 

Perhaps just another annotation indicating user requested drift so that Karpenter can replace nodes in an orderly manner and while respecting deprovisioning controls. 

**How important is this feature to you?**

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by njtran** [comment]
While you could just do `kubectl delete node`, it sounds like you want your own custom way to mark nodes for Karpenter to eventually deprovision them. We discussed doing this when migrating all deprovisioning into the deprovisioning controller as it is now, but we haven't seen the signal to implement something like this yet.

In the near future, we'll release some additional work needed for the full machine migration, which moves these disruption decisions to the Machine Status conditions here: https://github.com/aws/karpenter-core/pull/319. 

In all, I think it's reasonable to allow a custom controller to annotate nodes as needing deprovisioning, rather than letting users mess with the Expiration/Drift annotations.

May I ask what kind of custom signals you use to choose nodes as being deprovisioned? Is there a gap in the deprovisioning logic that we need to implement, or is it simply just a matter of filling in the gaps where Drift isn't implemented yet? 

If you're also willing to design and implement this (both should be relatively simple), I'm happy to walk through this with you and get you ramped up on it.

**Comment by sidewinder12s** [comment]
I mean, right now there are a ton of gaps. It was quite useful that I could just annotate all at once and Karpenter would do a serial replacement to minimize disruption in the cluster. It generally hasn't been clear to me how disruptive it'd be to just delete a pile of nodes wholesale.

Long term, I was sorta expecting there are always going to be gaps (is karpenter going to be able to drift on userdata changes for example?).

My current migration is dockerd -> containerd which is not drifted. 

I don't think I have slack probably this quarter to get into this.

**Comment by njtran** [comment]
> Long term, I was sorta expecting there are always going to be gaps (is karpenter going to be able to drift on userdata changes for example?).

Yeah we should be able to drift for all provisioning configurations in the future. This work is in-flight. Documentation [here] (https://karpenter.sh/docs/concepts/deprovisioning/#drift), and first PR of a couple [here](https://github.com/aws/karpenter-core/pull/400)

**Comment by njtran** [comment]
I updated the title to be more general based off our discussion. 

This could look like users adding an annotation/label/taint for Karpenter to discover and then implement it as [another deprovisioner](https://github.com/aws/karpenter-core/blob/main/pkg/controllers/deprovisioning/controller.go#L90-L103) in the Deprovisioning loop.

**Comment by sidewinder12s** [comment]
One workflow that has come up that might be applicable for this is how to eventually force all pods that are using `karpenter.sh/do-not-evict: "true"` to get consolidated/drifted in a timely manner.

We'd found that the nodes end up getting left around since Karpenter is also not cordoning the nodes that these pods are on to allow them to eventually drain out of the cluster. 

Another thought though would be to modify how this annotation is treated if maintenance window type support was added to Karpenter (ref: #1302). I think at least operationally allowing a maintenance window to force node replacements may be desirable by a lot of organizations.

**Comment by yuvalavidor** [comment]
> One workflow that has come up that might be applicable for this is how to eventually force all pods that are using karpenter.sh/do-not-evict: "true" to get consolidated/drifted in a timely manner.

I think that the obvious use case here is a cluster upgrade where you want to upgrade your nodes.
`kubectl delete node` is a way of doing that, but i would like to see a way that spins up a new node with the same size before the node is deleted to reduce the time for pods to be scheduled, and in a statefulset its even more time consuming

and also, im actually asking, what would be the best way of doing so in that situation right now (0.30.0)

**Comment by sidewinder12s** [comment]
> > One workflow that has come up that might be applicable for this is how to eventually force all pods that are using karpenter.sh/do-not-evict: "true" to get consolidated/drifted in a timely manner.
> 
> I think that the obvious use case here is a cluster upgrade where you want to upgrade your nodes. `kubectl delete node` is a way of doing that, but i would like to see a way that spins up a new node with the same size before the node is deleted to reduce the time for pods to be scheduled, and in a statefulset its even more time consuming
> 
> and also, im actually asking, what would be the best way of doing so in that situation right now (0.30.0)

I haven't really found any better way to deal with it than draining or deleting the node. I actually am not sure if with Karpenter now on delete would launch a new node before draining it.

**Comment by garvinp-stripe** [comment]
Wanted to add more to this. Currently we have an internal system that marks node for killing for reasons that K8s has no idea about for SOX and/or compliances or CVE related reasons. We are building a controller to read these and annotated our nodes for killing. We realize having 2 systems that kill nodes means it is difficult if not impossible to understand termination budget of the fleet so if we can get Karpenter to reap instead then we can maintain a realistic termination budget.

**Comment by njtran** [comment]
@garvinp-stripe or @sidewinder12s, let me know if you want to help implement this. I can help guide through this, and it should be fairly straightforward.
